rfcs/proposed/host_backend_histogram/README.md

```diff
@@ -0,0 +1,73 @@
+
+It remains to be seen if atomics are worth their overhead and contention from a performance perspective and may depend on the different approaches available.
+
+### Selecting Between Algorithms
+It may be the case that multiple aspects may provide an advantage to either algorithm one or two. Which `histogram` API has been called, `n`, the number of output bins, and backend/atomic provider may all impact the performance trade-offs between these two approaches. My intention is to experiment with these and be open to a heuristic to choose one or the other based upon the circumstances if that is what the data suggests is best. The larger the number of output bins, the better atomics should do vs redundant copies of the output.
+
+## Alternative Approaches
+* One could consider some sort of locking approach which locks mutexes for subsections of the output histogram prior to modifying them. It's possible such an approach could provide a similar approach to atomics, but with different overhead tradeoffs. It seems quite likely that this would result in more overhead, but it could be worth exploring.
+
+* Another possible approach could be to do something like the proposed implementation one, but with some sparse representation of output data. However, I think the general assumptions we can make about the normal case make this less likely to be beneficial. It is quite likely that `n` is much larger than the output histograms, and that a large percentage of the output histogram may be occupied, even when considering dividing the input amongst multiple threads. This could be explored if we find temporary storage is too large for some cases and the atomic approach does not provide a good fallback.
+
+## Open Questions
+* Would it be worthwhile to add our own implementation of `atomic_ref` for C++17? I believe this would require specializations for each of our supported compilers.
+
+* What is the overhead of atomics in general in this case and does the overhead there make them inherently worse than merely having extra copies of the histogram and accumulating?
+
+* Is it worthwhile to have separate implementations for TBB and OpenMP because they may differ in the best-performing implementation? What is the best heuristic for selecting between algorithms (if one is not the clear winner)?
+
+* How will vectorized bricks perform, and in what situations will it be advantageous to use or not use vector instructions?
```

Comment:
A couple of papers that might be of interest:
- https://www.researchgate.net/publication/221131224_SIMD_Vectorization_of_Histogram_Functions
- https://www.researchgate.net/publication/266660552_Versatile_and_scalable_parallel_histogram_construction

Reply:
Thanks, these are interesting, and mainly follow similar lines to the approaches outlined here.  I will look more into the `pcwar` SIMD instruction to understand if that is viable to use here.

Reply:
While its interesting to explore, I believe we basically depend on OpenMP simd to provide our SIMD operations.  We won't be able to serve more complex intrinsic operations if we want to stick to that.

Further, I'm not actually finding anything like the pcwar instruction(s) they refer to here in https://www.intel.com/content/www/us/en/docs/intrinsics-guide/index.html.  I do find some for conflict detection across lane which _could_ be useful, but again those won't be available through the interface of OpenMP.

I think our options are basically to decline SIMD or to have duplicates of the output for each SIMD lane. I think even that may be more "hands on" with SIMD details than we have done thus far from oneDPL.